wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,683 @@
+-- | Module providing restoring from backup phrase functionality
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeFamilies   #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Pos.Util.Mnemonic
+       (
+       -- * Types
+         Mnemonic
+       , Entropy
+       , SupportedWords
+
+       -- * Creating @Mnemonic@ (resp. @Entropy@)
+       , mkEntropy
+       , mkMnemonic
+       , genEntropy
+
+       -- * Access properties of types
+       , entropyLength
+       , checksumLength
+
+       -- * Converting from and to @Mnemonic@ (resp. @Entropy@)
+       , mnemonicToEntropy
+       , mnemonicToSeed
+       , mnemonicToAESKey
+       , entropyToMnemonic
+       , entropyToByteString
+       ) where
+
+import           Universum
+
+import           Control.Lens ((?~))
+import           Crypto.Hash (Blake2b_256, Digest, SHA256, hash)
+import           Data.Aeson (FromJSON (..), ToJSON (..))
+import           Data.Aeson.Types (Parser)
+import           Data.ByteArray (convert)
+import           Data.ByteString (ByteString)
+import           Data.Char (isAscii)
+import           Data.Default (Default (def))
+import           Data.List (elemIndex, (!!))
+import           Data.Swagger (NamedSchema (..), ToSchema (..), maxItems, minItems)
+import           Data.Text.Buildable (Buildable (build))
+import           GHC.TypeLits (ErrorMessage (..), TypeError)
+import           Test.QuickCheck (Arbitrary (..))
+import           Test.QuickCheck.Gen (vectorOf)
+
+import           Pos.Binary (serialize')
+import           Pos.Crypto (AesKey (..))
+import           Pos.Util.Bits (Bit, Word11, fromBits, toBits)
+import           Pos.Util.LogSafe (SecureLog)
+
+
+import qualified Crypto.Random.Entropy as Crypto
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as Text
+
+
+-- | A backup-phrase in the form of a non-empty of Mnemonic words
+-- Constructor isn't exposed.
+data Mnemonic (n :: Nat) = Mnemonic
+    { getEntropy :: Entropy n
+    , getWords   :: [(Word11, Text)]
+    } deriving (Eq, Show, Generic)
+
+-- | Entropy as a non-empty sequence of bytes, multiple of 4 bytes
+newtype Entropy (n :: Nat) = Entropy ByteString deriving (Eq, Show, Generic)
+
+type family SupportedWords (n :: Nat) :: Constraint where
+    SupportedWords 3  = ()
+    SupportedWords 6  = ()
```

Comment:
just to eliminate possible errors, I would rather remove 3 and 6 word mnemonics - as they don't provide enough entropy (to my knowledge) and even bip39 defines it should be 12 or more words (we used 9 for a specific corner use case)